iavl_node.go

```diff
@@ -312,49 +346,58 @@ func (node *IAVLNode) set(t *IAVLTree, key []byte, value []byte) (newSelf *IAVLN
 // newKey: new leftmost leaf key for tree after successfully removing 'key' if changed.
 // value: removed value.
 func (node *IAVLNode) remove(t *IAVLTree, key []byte) (
-	newHash []byte, newNode *IAVLNode, newKey []byte, value []byte, removed bool) {
+	newHash []byte, newNode *IAVLNode, newKey []byte, value []byte, orphaned []*IAVLNode,
+) {
 	if node.isLeaf() {
 		if bytes.Equal(key, node.key) {
-			removeOrphan(t, node)
-			return nil, nil, nil, node.value, true
-		} else {
-			return node.hash, node, nil, nil, false
+			return nil, nil, nil, node.value, []*IAVLNode{node}
 		}
+		return node.hash, node, nil, nil, []*IAVLNode{}
+	}
+
+	if bytes.Compare(key, node.key) < 0 {
+		var newLeftHash []byte
+		var newLeftNode *IAVLNode
+
+		newLeftHash, newLeftNode, newKey, value, orphaned =
+			node.getLeftNode(t).remove(t, key)
+
+		if len(orphaned) == 0 {
+			return node.hash, node, nil, value, []*IAVLNode{}
+		} else if newLeftHash == nil && newLeftNode == nil { // left node held value, was removed
+			return node.rightHash, node.rightNode, node.key, value, orphaned
+		}
+		orphaned = append(orphaned, node)
+
+		node = node.clone()
+		node.leftHash, node.leftNode = newLeftHash, newLeftNode
```

Comment:
Do not see where you're setting the node.key to newKey for the left node. https://github.com/tendermint/iavl/pull/1/files#diff-a5ef36e3990bde5c8239a4c18ec3d25dR395

Reply:
Hmm, so this code was adapted from the old version of iavl, which has that peculiarity too, if you look at the deleted code. I'll take a deeper look, but didn't want to break anything by changing the logic. I'm not sure why it works this way, and whether or not it's correct. I've also tried removing that `if` statement, and the tests still pass :grimacing:

Reply:
LOL, ok